### PR TITLE
Profiles: fix New profile with current pdf document settings

### PR DESCRIPTION
**Pull request by hius07**
Didn't work for non-English languages.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10778)

### CODE REVIEW

File: plugins/profiles.koplugin/main.lua [hunk 1]
```diff
@@ -337,10 +337,19 @@ function Profiles:getProfileFromCurrentDocument(new_name)
...
+    local setting_needs_arg = {
+        ["kopt_trim_page"]       = true,
+        ["kopt_zoom_mode_genus"] = true,
+        ["kopt_zoom_mode_type"]  = true,
+    }
 
     local profile = { settings = { name = new_name, order = document_settings } }
     for _, v in ipairs(document_settings) do
-        profile[v] = self.document.configurable[self.ui.rolling and v or v:sub(6)]
+        local value = self.document.configurable[self.ui.rolling and v or v:sub(6)]
```
Comment (poire-z) [inline plugins/profiles.koplugin/main.lua:348]: Looking at the code around, I couldn't find the answer: what are these 5 first chars removed by this `v:sub(6)` ?
I guess it's `kopt_` (looking at one of my metadata.pdf.lua) ? Might be worth a comment around here.
Btw, no issue in there with profiles with your cre/rolliing settings deduplication and the main and unique setting now being `copt_something`?

**Comment by hius07** [comment]
Dispatcher internally stores "copt configurable" actions as is, "kopt configurable" actions with the prefix:
https://github.com/koreader/koreader/blob/2ba3353d8a356234ee8343475ef7c0fa27e9b047/frontend/dispatcher.lua#L486-L491
Following that, `Profiles:getProfileFromCurrentDocument()` gives the prefix to the pdf options.

The document "configurable" settings do not have the prefix, so it should be stripped when getting a setting value.
Deduplication doesn't affect this, we get the value from the memory, not from the file.

File: frontend/dispatcher.lua [hunk 2]
```diff
@@ -584,6 +584,11 @@ function Dispatcher:getNameFromItem(item, settings, dont_show_value)
     return title
 end
 
+function Dispatcher:getArgFromValue(item, value)
```
Comment (yparitcher) [inline frontend/dispatcher.lua:587]: On mobile, so i can't code search too well, but is this trying to reverse from the display string to the argument?

Do we store the translated string for the setting?

Comment (hius07) [reply frontend/dispatcher.lua:587]: https://github.com/koreader/koreader/blob/ccd29d36c0cde78d956bbd01b7232a9d70c81ffd/frontend/ui/data/koptoptions.lua#L86-L92
For exmple, `trim_page`.
It converts `values` (that is stored in `document.configurable.trim_page`) to `args` (that is expected by Dispatcher).

Comment (hius07) [reply frontend/dispatcher.lua:587]: Yes, the translated string (from `toggle`) is displayed in QuickMenu.

File: plugins/profiles.koplugin/main.lua [hunk 3]
```diff
@@ -350,13 +360,6 @@ function Profiles:getProfileFromCurrentDocument(new_name)
         profile["embedded_fonts"] = self.ui.typeset.embedded_fonts
         profile["smooth_scaling"] = self.ui.typeset.smooth_scaling
     else
-        local trim_page_to_mode = { _("manual"), _("auto"), _("semi-auto"), _("none") }
-        local zoom_genus_to_mode = { _("manual"), _("rows"), _("columns"), _("content"), _("page") }
-        local zoom_type_to_mode = { _("height"), _("width"), _("full") }
```
Comment (yparitcher) [inline plugins/profiles.koplugin/main.lua:355]: Did you try this without the translation?
Like:
```
local trim_page_to_mode = { "manual", "auto", "semi-auto", "none"}
```

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: That's what I started from.
But it's a duplication of data that we usually try to avoid.

Comment (yparitcher) [reply plugins/profiles.koplugin/main.lua:355]: For zooming you can use
```
ReaderZooming.zoom_mode_to_genus
ReaderZooming.zoom_mode_to_type
```

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: Isn't it better to deduplicate ReaderZooming as well and have all that strings in one storage only - koptoptions?
Dispatcher would provide values/args/labels conversion when needed.
@NiLuJe ?

Comment (poire-z) [reply plugins/profiles.koplugin/main.lua:355]: Will this survive a change of UI language ? if you store (dunno if you do) such translated strings?

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: In profile.lua I should not store translatable strings, it was a mistake.
In QM, Dispatcher shows translated "labels" correctly.

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: > Isn't it better to deduplicate ReaderZooming as well and have all that strings in one storage only - koptoptions? Dispatcher would provide values/args/labels conversion when needed. @NiLuJe ?

I kinda lost the plot a bit, what are we talking about exactly?

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: For creating a new profile with current pdf document settings we need arguments for zoom_mode settings.
Configurable stores values only.
I propose to ask koptoptions (from Dispatcher) to get arg from value.
Other way is to duplicate the value/arg table from koptoptions to profiles.lua.
You did so in https://github.com/koreader/koreader/pull/7463 (duplicated values/args tables).
I propose to think about the total deduplication and have one storage of the settings - koptoptions.
Dispatcher would provide values/args/labels conversion when needed.

Comment (Frenzie) [reply plugins/profiles.koplugin/main.lua:355]: I don't necessarily know if it makes sense to make the strings in Dispatcher identical to those in the bottom panel or if a context should be used, but we haven't had any complaints yet afaik.

Comment (Frenzie) [reply plugins/profiles.koplugin/main.lua:355]: That is to say, total deduplication could be sensible or it could be trying to be a bit too smart and I'm not quite sure which it is. :-)

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: Unless I missed something (and I very well might, ReaderZooming is nightmare fuel, and I'm not familiar with ConfigDialog), but RZ itself is already using that table, so it's really duplication per-se. If the LUT is already used somewhere, no need to complicate matters more.

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: In fact, RZ settings are so brittle that I'd vote *not* to rock the boat too much... On the other hand, that's also a very good reason to think about doing it better ;o).

TL;DR: If you feel like it, do what you feel is best.

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: > if it makes sense to make the strings in Dispatcher identical to those in the bottom panel

They are not identical but just the same, both ConfigDialog and Dispatcher read them from coptoptions/koptoptions.

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: > RZ itself is already using that table

Example of duplication:
https://github.com/koreader/koreader/blob/493a60faefbee6762b96600e6025ea15dc050c51/frontend/ui/data/koptoptions.lua#L285
https://github.com/koreader/koreader/blob/493a60faefbee6762b96600e6025ea15dc050c51/frontend/apps/reader/modules/readerzooming.lua#L40-L46
I proposed a new function `Dispatcher:getArgFromValue()`.

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: The fact that page matches to 1 in the first case and 4 in the latter would make me want to keep this distinct to make that quirk very clear, but, again, not sure if relevant here ;).

(Also, I slept badly tonight, so I'm sorry if I'm slow and/or feel particularly unhelpful today ^^).

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: > page matches to 1 in the first case and 4 in the latter

It's okay
https://github.com/koreader/koreader/blob/493a60faefbee6762b96600e6025ea15dc050c51/frontend/ui/data/koptoptions.lua#L281

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: I don't really want to lose the RZ LUT (because RZ itself uses it, and the RZ settings are such a mess that having it like that in there helps readability immensely), regardless of what KoptOptions does, FWIW.

So, eeeh :/.

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:355]: I'm also not terribly fond of the idea of *Options or ConfigDialog being the source of truth, too, FWIW. It makes working with individual modules much murkier, i'd much rather each modules keep being the source of truth for their own stuff.

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:355]: Got it, thanks.

Comment (yparitcher) [reply plugins/profiles.koplugin/main.lua:355]: Ok, i understand now :)

Please put a comment explaining what the function does, something like: "converts koptoptions values to args"

**Review by yparitcher — APPROVED** [review]
